bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/PredefinedTypes.java

```diff
@@ -170,9 +170,15 @@
 
     public static final RecordType STRING_ITR_NEXT_RETURN_TYPE =
             IteratorUtils.createIteratorNextReturnType(PredefinedTypes.TYPE_STRING);
-    public static final RecordType XML_ITR_NEXT_RETURN_TYPE = IteratorUtils
-            .createIteratorNextReturnType(
-                    new BUnionType(Arrays.asList(PredefinedTypes.TYPE_STRING, PredefinedTypes.TYPE_XML)));
+
+    public static final RecordType XML_ITR_NEXT_RETURN_ELEMENT_TYPE =
+            IteratorUtils.createIteratorNextReturnType(TYPE_ELEMENT);
+    public static final RecordType XML_ITR_NEXT_RETURN_TEXT_TYPE =
+            IteratorUtils.createIteratorNextReturnType(TYPE_TEXT);
```

Comment:
How come we don't need to do this for processing instruction and comment sequences?

Reply:
They are not iterable so I didn't not create iteration return types for them.

Reply:
> They are not iterable so I didn't not create iteration return types or write type param logic for them.

`xml:Element` also isn't iterable, right? But there's a `XML_ITR_NEXT_RETURN_ELEMENT_TYPE`?

These are the record value types for iterator next right?

We can iterate `xml`, `xml<T>`, and `xml:Text`. 

So I assume we will need record values with all of the following field types
- `xml:Element|xml:Text|xml:ProcessingInstruction|xml:Comment`
- `xml:Element`
- `xml:Text`
- `xml:ProcessingInstruction`
- `xml:Comment`
- and other combinations of unions of these (e.g., `xml:Comment|xml:Element` for `xml<xml:Comment|xml:Element>`).

Thus my question re: why we don't need the others. How do we handle the union case btw?

Reply:
xml<xml:Text> and xml<xml:ProcessingInstruction> are not iterable in the sense it returns null from `next`

for examples like:
```
xml d = xml `<!--first-->`;
        xml<'xml:Element|'xml:Comment> seq4 = <xml<'xml:Element|'xml:Comment>> d.concat(xml `<foo>fu</foo>`);

         int i =0;
         foreach 'xml:Element|'xml:Comment item in seq4 {
             concatIntString(i, item.toString());
             i += 1;
         }

         record {| 'xml:Element|'xml:Comment value; |}? nextEle = seq4.iterator().next();

         panic error("out for loop: "+ output + "\n out record: "+ nextEle.toString());
```
the current implementation returns: 
```
error: out for loop: 0:<!--first--> 1:<foo>fu</foo> 
 out record: {"value":`<!--first-->`}
```
This is the expected output right?

The BXml type for this is : 

> xml<lang.xml:Element|lang.xml:Comment|lang.xml:ProcessingInstruction|lang.xml:Text>

I'm switching from the  BXml type.


`XML_ITR_NEXT_RETURN_ELEMENT_TYPE` is to handle examples like:

```
xml xdata = xml `<p:person xmlns:p="foo" xmlns:q="bar">
                <p:name>bob</p:name>
                <p:address>
                    <p:city>NY</p:city>
                    <q:country>US</q:country>
                </p:address>
                <q:ID>1131313</q:ID>
            </p:person>`;
```

who's BXml type is `lang.xml:Element`

Reply:
> xmlxml:Text and xmlxml:ProcessingInstruction are not iterable in the sense it returns null from `next`
> 

IINM
- `xml<'xml:Text>` is the same as `'xml:Text` and iteration will always return one value which is the `'xml:Text` itself
- Iterating `xml<'xml:ProcessingInstruction>` will return 0 or more `'xml:ProcessingInstruction`s

I'm not sure I get what you mean by `next` returning null. 

> for examples like:
> 
> ```
> xml d = xml `<!--first-->`;
>         xml<'xml:Element|'xml:Comment> seq4 = <xml<'xml:Element|'xml:Comment>> d.concat(xml `<foo>fu</foo>`);
> 
>          int i =0;
>          foreach 'xml:Element|'xml:Comment item in seq4 {
>              concatIntString(i, item.toString());
>              i += 1;
>          }
> 
>          record {| 'xml:Element|'xml:Comment value; |}? nextEle = seq4.iterator().next();
> 
>          panic error("out for loop: "+ output + "\n out record: "+ nextEle.toString());
> ```
> 
> the current implementation returns:
> 
> ```
> error: out for loop: 0:<!--first--> 1:<foo>fu</foo> 
>  out record: {"value":`<!--first-->`}
> ```
> 
> This is the expected output right?
> 
> The BXml type for this is :
> 
> > xmllang.xml:Element|lang.xml:Comment|lang.xml:ProcessingInstruction|lang.xml:Text

The output is correct, but shouldn't the iteration value type here be `'xml:Element|'xml:Comment` since we're iterating over `xml<'xml:Element|'xml:Comment>`?

> 
> I'm switching from the BXml type.
> 
> `XML_ITR_NEXT_RETURN_ELEMENT_TYPE` is to handle examples like:
> 
> ```
> xml xdata = xml `<p:person xmlns:p="foo" xmlns:q="bar">
>                 <p:name>bob</p:name>
>                 <p:address>
>                     <p:city>NY</p:city>
>                     <q:country>US</q:country>
>                 </p:address>
>                 <q:ID>1131313</q:ID>
>             </p:person>`;
> ```
> 
> who's BXml type is `lang.xml:Element`

Not sure why we need to special case this. Again, IMO the iteration value type should be based on the variable's type, and not the type of the actual value it holds.

When we iterate over `xdata`, we're iterating over an `xml` which is `xml<'xml:Element|'xml:Comment|'xml:Text|'xml:ProcessingInstruction>`, so IMO the iteration type here should be `'xml:Element|'xml:Comment|'xml:Text|'xml:ProcessingInstruction`.

Reply:
As we discussed `xml<xml:Comment> and xml<xml:ProcessingInstruction>` is now iterable. Also regarding unions, we are unable to write a xml union sequence with lang capabilities at the moment. However, after discussing with @rdhananjaya,  instead of switching from 'BXml' types it is now switched from the 'xml value type' so a union case won't exist now.